sonar-plugin-api/src/main/java/org/sonar/api/web/ServletFilter.java

```diff
@@ -50,141 +49,100 @@ public UrlPattern doGetPattern() {
 
   public static final class UrlPattern {
 
-    private static final String MATCH_ALL_URL = "/*";
+    private static final String MATCH_ALL = "/*";
 
-    private final boolean matchAllUrls;
-    private final List<Pattern> includePatterns = new ArrayList<>();
-    private final List<Pattern> excludePatterns = new ArrayList<>();
+    private final Predicate<String>[] inclusions;
+    private final Predicate<String>[] exclusions;
 
     private UrlPattern(Builder builder) {
-      this.includePatterns.addAll(builder.includePatterns.stream().map(Pattern::new).collect(Collectors.toList()));
-      this.excludePatterns.addAll(builder.excludePatterns.stream().map(Pattern::new).collect(Collectors.toList()));
-      this.matchAllUrls = excludePatterns.isEmpty() && includePatterns.size() == 1 && includePatterns.get(0).getUrl().equals(MATCH_ALL_URL);
+      if (builder.inclusions.isEmpty()) {
+        // because Stream#anyMatch() returns false if stream is empty
+        this.inclusions = new Predicate[] {s -> true};
+      } else {
+        this.inclusions = builder.inclusions.stream().toArray(Predicate[]::new);
+      }
+      this.exclusions = builder.exclusions.stream().toArray(Predicate[]::new);
     }
 
     public boolean matches(String path) {
-      // Optimization for filter that match all urls
-      return matchAllUrls ||
-      // Otherwise we first verify if the url is not excluded, then we verify if the url is included
-        (!excludePatterns.stream().anyMatch(pattern -> pattern.matches(path)) && includePatterns.stream().anyMatch(pattern -> pattern.matches(path)));
-    }
-
-    /**
-     * @since 6.0
-     */
-    public Set<String> getIncludePatterns() {
-      return includePatterns.stream().map(Pattern::getUrl).collect(Collectors.toSet());
-    }
-
-    /**
-     * @since 6.0
-     */
-    public Set<String> getExcludePatterns() {
-      return excludePatterns.stream().map(Pattern::getUrl).collect(Collectors.toSet());
+      return !Arrays.stream(exclusions).anyMatch(pattern -> pattern.test(path)) &&
+        Arrays.stream(inclusions).anyMatch(pattern -> pattern.test(path));
     }
 
     /**
-     * @deprecated since 6.0, use {@link #getIncludePatterns()} or {@link #getExcludePatterns()} instead
+     * @return always return empty string {@code ""} since version 6.0
+     * @deprecated since 6.0 as multiple inclusion and exclusion patterns can be defined
      */
     @Deprecated
-    public String getUrl(){
-      // Before 6.0, it was only possible to include one url
-      if (excludePatterns.isEmpty() && includePatterns.size() == 1 ) {
-        return includePatterns.get(0).getUrl();
-      }
-      throw new IllegalStateException("this method is deprecated and should not be used anymore");
+    public String getUrl() {
+      return "";
```

Comment:
As discussed, you should return value if only one included pattern exists in order to not break backward compatibility